src/Microsoft.AspNetCore.Mvc.TagHelpers/CacheTagKey.cs

```diff
@@ -0,0 +1,428 @@
+        private IList<KeyValuePair<string, string>> _routeValues;
+        private IList<KeyValuePair<string, string>> _cookies;
+        private bool _varyByUser;
+        private string _username;
+
+        private string _generatedKey;
+        private int? _hashcode;
+        
+        private CacheTagKey()
+        {
+        }
+
+        public static CacheTagKey From(CacheTagHelper tagHelper, TagHelperContext context)
+        {
+            var httpContext = tagHelper.ViewContext.HttpContext;
+            var request = httpContext.Request;
+
+            var cacheKey = new CacheTagKey();
+
+            cacheKey._key = context.UniqueId;
+            cacheKey._prefix = nameof(CacheTagHelper);
+
+            cacheKey._expiresAfter = tagHelper.ExpiresAfter;
+            cacheKey._expiresOn = tagHelper.ExpiresOn;
+            cacheKey._expiresSliding = tagHelper.ExpiresSliding;
+            cacheKey._varyBy = tagHelper.VaryBy;
+            cacheKey._cookies = ExtractCookies(tagHelper.VaryByCookie, request.Cookies);
+            cacheKey._headers = ExtractHeaders(tagHelper.VaryByHeader, request.Headers);
+            cacheKey._queries = ExtractQueries(tagHelper.VaryByQuery, request.Query);
+            cacheKey._routeValues = ExtractRoutes(tagHelper.VaryByRoute, tagHelper.ViewContext.RouteData.Values);
+            cacheKey._varyByUser = tagHelper.VaryByUser;
+
+            if (cacheKey._varyByUser)
+            {
+                cacheKey._username = httpContext.User?.Identity?.Name;
+            }
+
+            return cacheKey;
+        }
+
+        public static CacheTagKey From(DistributedCacheTagHelper tagHelper, TagHelperContext context)
+        {
+            var httpContext = tagHelper.ViewContext.HttpContext;
+            var request = httpContext.Request;
+
+            var cacheKey = new CacheTagKey();
+
+            cacheKey._key = tagHelper.Name;
+            cacheKey._prefix = nameof(DistributedCacheTagHelper);
+
+            cacheKey._expiresAfter = tagHelper.ExpiresAfter;
+            cacheKey._expiresOn = tagHelper.ExpiresOn;
+            cacheKey._expiresSliding = tagHelper.ExpiresSliding;
+            cacheKey._varyBy = tagHelper.VaryBy;
+            cacheKey._cookies = ExtractCookies(tagHelper.VaryByCookie, request.Cookies);
+            cacheKey._headers = ExtractHeaders(tagHelper.VaryByHeader, request.Headers);
+            cacheKey._queries = ExtractQueries(tagHelper.VaryByQuery, request.Query);
+            cacheKey._routeValues = ExtractRoutes(tagHelper.VaryByRoute, tagHelper.ViewContext.RouteData.Values);
+            cacheKey._varyByUser = tagHelper.VaryByUser;
+
+            if (cacheKey._varyByUser)
+            {
+                cacheKey._username = httpContext.User?.Identity?.Name;
+            }
+
+            return cacheKey;
+        }
+
+        private static IList<KeyValuePair<string, string>> ExtractCookies(string keys, IRequestCookieCollection cookies)
```

Comment:
Instead of duplicating each ofthese methods you should go back to the single method with an accessor lambda. To make sure you don't allocate each time you can hoist the lambda.